storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -575,38 +683,75 @@ func (s *commitLogSource) ReadIndex(
 		resultOptions  = s.opts.ResultOptions()
 	)
 
+	// Start by reading all the commit log files that we couldn't eliminate due to the
+	// cached metadata.
 	for iter.Next() {
 		series, dp, _, _ := iter.Current()
-		if !s.shouldIncludeInIndex(
-			series.Shard, dp.Timestamp, highestShard, indexBlockSize, bootstrapRangesByShard) {
-			continue
-		}
 
-		segment, err := indexResults.GetOrAddSegment(dp.Timestamp, indexOptions, resultOptions)
-		if err != nil {
-			return nil, err
-		}
+		s.maybeAddToIndex(
+			series.ID, series.Tags, series.Shard, highestShard, dp.Timestamp, bootstrapRangesByShard,
+			indexResults, indexOptions, indexBlockSize, resultOptions)
+	}
 
-		exists, err := segment.ContainsID(series.ID.Data().Bytes())
-		if err != nil {
-			return nil, err
-		}
-		if exists {
-			continue
+	// Add in all the data that was cached by a previous run of ReadData() (if any).
+	for shard, shardData := range cachedShardData {
+		for _, series := range shardData.series {
+			for dataBlockStart := range series.encoders {
+				s.maybeAddToIndex(
+					series.id, series.tags, uint32(shard), highestShard, dataBlockStart.ToTime(), bootstrapRangesByShard,
+					indexResults, indexOptions, indexBlockSize, resultOptions)
+			}
 		}
+	}
 
-		d, err := convert.FromMetric(series.ID, series.Tags)
-		if err != nil {
-			return nil, err
-		}
+	return indexResult, nil
+}
 
-		_, err = segment.Insert(d)
-		if err != nil {
-			return nil, err
-		}
+func (s commitLogSource) maybeAddToIndex(
+	id ident.ID,
+	tags ident.Tags,
+	shard uint32,
+	highestShard uint32,
+	blockStart time.Time,
+	bootstrapRangesByShard []xtime.Ranges,
+	indexResults result.IndexResults,
+	indexOptions namespace.IndexOptions,
+	indexBlockSize time.Duration,
+	resultOptions result.Options,
+) error {
+	if !s.shouldIncludeInIndex(
+		shard, blockStart, highestShard, indexBlockSize, bootstrapRangesByShard) {
+		return nil
 	}
 
-	return indexResult, nil
+	segment, err := indexResults.GetOrAddSegment(blockStart, indexOptions, resultOptions)
+	if err != nil {
+		return err
+	}
+
+	exists, err := segment.ContainsID(id.Data().Bytes())
```

Comment:
You'll have a conflict, make sure to rebase on master (`Data().Bytes()` no longer exists and you need to do just `Bytes()`).